internal/watcher/health/health_watcher_service.go

```diff
@@ -15,57 +16,67 @@ import (
 	"github.com/nginx/agent/v3/internal/config"
 	"github.com/nginx/agent/v3/internal/logger"
 
-	"github.com/nginx/agent/v3/api/grpc/mpi/v1"
+	mpi "github.com/nginx/agent/v3/api/grpc/mpi/v1"
 )
 
 //go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
 //counterfeiter:generate . healthWatcherOperator
 
 type (
 	healthWatcherOperator interface {
-		Health(ctx context.Context, instanceID string) *v1.InstanceHealth
+		Health(ctx context.Context, instance *mpi.Instance) (*mpi.InstanceHealth, error)
 	}
 
 	HealthWatcherService struct {
 		agentConfig *config.Config
-		cache       map[string]*v1.InstanceHealth    // key is instanceID
+		cache       map[string]*mpi.InstanceHealth   // key is instanceID
 		watchers    map[string]healthWatcherOperator // key is instanceID
+		instances   map[string]*mpi.Instance         // key is instanceID
 	}
 
 	InstanceHealthMessage struct {
-		correlationID  slog.Attr
-		instanceHealth []*v1.InstanceHealth
+		CorrelationID  slog.Attr
+		InstanceHealth []*mpi.InstanceHealth
 	}
 )
 
 func NewHealthWatcherService(agentConfig *config.Config) *HealthWatcherService {
 	return &HealthWatcherService{
 		watchers:    make(map[string]healthWatcherOperator),
-		cache:       make(map[string]*v1.InstanceHealth),
+		cache:       make(map[string]*mpi.InstanceHealth),
+		instances:   make(map[string]*mpi.Instance),
 		agentConfig: agentConfig,
 	}
 }
 
-func (hw *HealthWatcherService) AddHealthWatcher(instances []*v1.Instance) {
+func (hw *HealthWatcherService) AddHealthWatcher(instances []*mpi.Instance) {
 	for _, instance := range instances {
 		switch instance.GetInstanceMeta().GetInstanceType() {
-		case v1.InstanceMeta_INSTANCE_TYPE_NGINX, v1.InstanceMeta_INSTANCE_TYPE_NGINX_PLUS:
+		case mpi.InstanceMeta_INSTANCE_TYPE_NGINX, mpi.InstanceMeta_INSTANCE_TYPE_NGINX_PLUS:
 			watcher := NewNginxHealthWatcher()
 			hw.watchers[instance.GetInstanceMeta().GetInstanceId()] = watcher
-		case v1.InstanceMeta_INSTANCE_TYPE_AGENT:
-		case v1.InstanceMeta_INSTANCE_TYPE_UNSPECIFIED,
-			v1.InstanceMeta_INSTANCE_TYPE_UNIT:
+		case mpi.InstanceMeta_INSTANCE_TYPE_AGENT:
+		case mpi.InstanceMeta_INSTANCE_TYPE_UNSPECIFIED,
+			mpi.InstanceMeta_INSTANCE_TYPE_UNIT:
 			fallthrough
 		default:
 			slog.Warn("Health watcher not implemented", "instance_type",
 				instance.GetInstanceMeta().GetInstanceType())
 		}
+		hw.instances[instance.GetInstanceMeta().GetInstanceId()] = instance
 	}
 }
 
-func (hw *HealthWatcherService) DeleteHealthWatcher(instances []*v1.Instance) {
+func (hw *HealthWatcherService) UpdateHealthWatcher(instances []*mpi.Instance) {
+	for _, instance := range instances {
+		hw.instances[instance.GetInstanceMeta().GetInstanceId()] = instance
+	}
+}
+
+func (hw *HealthWatcherService) DeleteHealthWatcher(instances []*mpi.Instance) {
 	for _, instance := range instances {
 		delete(hw.watchers, instance.GetInstanceMeta().GetInstanceId())
+		delete(hw.instances, instance.GetInstanceMeta().GetInstanceId())
 	}
 }
```

Comment:
Shouldn't we always delete the `cache` entry here for the Instance too? That way we wouldn't need to delete `cache`entries in `updateCache` 🤔

Reply:
Discussed in office, the instance isn't deleted in `DeleteHealthWatcher()` as we need a way to tell if an instance has been deleted in `health()` so we can send a health status for that instance that is unhealthy. This is currently being done by comparing the cache with the instanceList. 
Add comments to updateCache to explain what the function is doing